Optimizer for aggregations with distinct mask

stagraqubole opened the pull request:
Optimizes queries which have aggregations on both distinct mask and other symbols

E.g.: 

```
presto:default> explain select o_custkey, COUNT(o_shippriority), sum(DISTINCT o_orderkey) FROM orders group by o_custkey;
                                                                                                                                  Query Plan                                                                                                                              
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[o_custkey, _col1, _col2] => [o_custkey:integer, count:bigint, sum:bigint]                                                                                                                                                                                       
         _col1 := count                                                                                                                                                                                                                                                   
         _col2 := sum                                                                                                                                                                                                                                                     
     - RemoteExchange[GATHER] => o_custkey:integer, count:bigint, sum:bigint                                                                                                                                                                                              
         - Project => [o_custkey:integer, count:bigint, sum:bigint]                                                                                                                                                                                                       
             - Aggregate(FINAL)[o_custkey] => [o_custkey:integer, $hashvalue:bigint, count:bigint, sum:bigint]                                                                                                                                                            
                     count := "arbitrary"("arbitrary")                                                                                                                                                                                                                    
                     sum := "sum"("sum_18")                                                                                                                                                                                                                               
                 - RemoteExchange[REPARTITION] => o_custkey:integer, arbitrary:bigint, sum_18:bigint, $hashvalue:bigint                                                                                                                                                   
                     - Aggregate(PARTIAL)[o_custkey] => [o_custkey:integer, $hashvalue_21:bigint, arbitrary:bigint, sum_18:bigint]                                                                                                                                        
                             arbitrary := "arbitrary"("expr_16")                                                                                                                                                                                                          
                             sum_18 := "sum"("expr_15")                                                                                                                                                                                                                   
                         - Project => [o_custkey:integer, $hashvalue_21:bigint, expr_15:bigint, expr_16:bigint]                                                                                                                                                           
                                 $hashvalue_21 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("o_custkey"), 0))                                                                                                                                             
                                 expr_15 := (CASE WHEN ("group" = CAST(1 AS BIGINT)) THEN "o_orderkey_2" END)                                                                                                                                                             
                                 expr_16 := (CASE WHEN ("group" = CAST(0 AS BIGINT)) THEN "count_14" END)                                                                                                                                                                 
                             - Aggregate(FINAL)[o_custkey, o_orderkey_2, group] => [o_custkey:integer, o_orderkey_2:bigint, group:bigint, $hashvalue_19:bigint, count_14:bigint]                                                                                          
                                     count_14 := "count"("count_17")                                                                                                                                                                                                      
                                 - RemoteExchange[REPARTITION] => o_custkey:integer, o_orderkey_2:bigint, group:bigint, count_17:bigint, $hashvalue_19:bigint                                                                                                             
                                     - Aggregate(PARTIAL)[o_custkey, o_orderkey_2, group] => [o_custkey:integer, o_orderkey_2:bigint, group:bigint, $hashvalue_20:bigint, count_17:bigint]                                                                                
                                             count_17 := "count"("o_shippriority")                                                                                                                                                                                        
                                         - Project => [o_custkey:integer, o_shippriority:integer, o_orderkey_2:bigint, group:bigint, $hashvalue_20:bigint]                                                                                                                
                                                 $hashvalue_20 := "combine_hash"("combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("o_custkey"), 0)), COALESCE("$operator$hash_code"("o_orderkey_2"), 0)), COALESCE("$operator$hash_code"("group"),
                                             - GroupId[[o_custkey, o_shippriority], [o_custkey, o_orderkey_2]] => [o_custkey:integer, o_shippriority:integer, o_orderkey_2:bigint, group:bigint]                                                                          
                                                 - Project => [o_shippriority:integer, o_orderkey_2:bigint, o_custkey:integer]                                                                                                                                            
                                                         o_orderkey_2 := CAST("o_orderkey" AS BIGINT)                                                                                                                                                                     
                                                     - TableScan[hive:hive:default:orders, originalConstraint = true] => [o_orderkey:integer, o_custkey:integer, o_shippriority:integer]                                                                                  
                                                             LAYOUT: hive                                                                                                                                                                                                 
                                                             o_orderkey := HiveColumnHandle{clientId=hive, name=o_orderkey, hiveType=int, hiveColumnIndex=0, columnType=REGULAR}                                                                                          
                                                             o_custkey := HiveColumnHandle{clientId=hive, name=o_custkey, hiveType=int, hiveColumnIndex=1, columnType=REGULAR}                                                                                            
                                                             o_shippriority := HiveColumnHandle{clientId=hive, name=o_shippriority, hiveType=int, hiveColumnIndex=7, columnType=REGULAR}     
```

stagraqubole commented:
Checked performance with this query:
```
SELECT ss_store_sk, 
       Sum(ss_quantity)               AS s, 
       Count(DISTINCT ss_customer_sk) AS d 
FROM   store_sales_orc_100
GROUP  BY ss_store_sk
```

With optimization, it works 2x-2.5x faster.

This optimizer can be extended to handle multiple distinct aggregations on different masks as well.

maciejgrzybek commented:
@stagraqubole thank you for this contribution!
Can you add some DSL testing (https://github.com/prestodb/presto/pull/4404)? Example of use: https://github.com/prestodb/presto/pull/5714/commits/27de6dbe9af62b5e21166822bc1e7e76534a6057

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java:
```diff
@@ -0,0 +1,436 @@
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+    private static final String HASH_CODE = FunctionRegistry.mangleOperatorName("HASH_CODE");
```
Not used

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java:
```diff
@@ -0,0 +1,436 @@
+            Optional<AggregateInfo> aggregateInfo = context.get();
+            // presence of aggregateInfo => mask also present
+            if (aggregateInfo.isPresent() && aggregateInfo.get().getMask().get().equals(node.getMarkerSymbol())) {
+                PlanNode source = node.getSource();
```
Call rewrite on source?

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java:
```diff
@@ -0,0 +1,436 @@
+    private final Metadata metadata;
+    private static final String HASH_CODE = FunctionRegistry.mangleOperatorName("HASH_CODE");
+
+    public DistinctOptimizer(Metadata metadata)
```
Maybe give this class more meanigful name. `ExtractDistinctAggregationOptimizer`?

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java:
```diff
@@ -0,0 +1,436 @@
+        }
+
+        @Override
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<AggregateInfo>> context)
```
If you code extract parts of this method to other methods - it could make it more readable. It is very long and deep.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java:
```diff
@@ -0,0 +1,436 @@
+                // symbols present in Group_i will be set, rest will be Null
+
+                //g0
+                if (!nonDistinctAggregateSymbols.isEmpty()) {
```
Is that possible provided the check `(masks.size() != 1 || node.getMasks().size() == node.getAggregations().size())` you do before.

stagraqubole replied:
You are right, wont happen right now. This is there because I was thinking of extending this to multiple distinct aggregations. Will remove it for now.

losipiuk commented:
Could you explain why this optimization improves performance?
Maybe this is good as short term solution but actually we should improve performance of execution part which is responsible for distinct aggregations.

stagraqubole commented:
@maciejgrzybek @losipiuk thanks for the comments, I will resolve them.

Regarding performance, improvement is because of the lesser data transferred. With the MarkDistinct plan, we transfer all data from SOURCE stage to next, with this optimization we are are partially grouping by the distinct symbol at SOURCE stage and sending data.

kokosing commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -61,6 +61,7 @@
     public static final String SPLIT_CONCURRENCY_ADJUSTMENT_INTERVAL = "split_concurrency_adjustment_interval";
     public static final String OPTIMIZE_METADATA_QUERIES = "optimize_metadata_queries";
     public static final String QUERY_PRIORITY = "query_priority";
+    public static final String OPTIMIZE_DISTINCT_AGGREGATIONS = "optimize_distinct_aggregations";
```
Please extract adding session property to separate commit

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java:
```diff
@@ -0,0 +1,436 @@
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        if (SystemSessionProperties.isOptimizeDistinctAggregationEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Optimizer(idAllocator, symbolAllocator, metadata),
```
static import

kokosing commented:
I have just skimmed, I will look more closely tomorrow.

kokosing commented:
@stagraqubole back to you

sopel39 commented:
This is an interesting optimization. Another idea would be rewrite `MarkDistinct` into `AggregationOperator` that would compute original `F1,...Fm`. It would also compute `c` distinct values partially at `SOURCE` stage and apply `F` on distinct `c` in final aggregation stage.
Such approach would reduce processing overhead and inter node pages copying between operators (e.g: overall plan would be simpler).

stagraqubole commented:
@sopel39 This in a way does that as it groups by mask symbol in first aggregation.

I have made changes as per review comments:
- Added DSL tests
- Changed optimizer name and refactored code a bit
- Added separate commit for session property

@kokosing @losipiuk @maciejgrzybek ready for another round of review

martint commented:
@erichwang, can you review?

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+
+        public List<Symbol> getOriginalNonDistinctAggregateArgs()
+        {
+            Set<Symbol> arguments = new HashSet<Symbol>();
```
You can write it like this, instead:

```java
return aggregations.values().stream()
        .filter(function -> !function.isDistinct())
        .flatMap(function -> function.getArguments().stream())
        .distinct()
        .map(Symbol::from)
        .collect(Collectors.toList());
```

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+
+        public List<Symbol> getOriginalDistinctAggregateArgs()
+        {
+            Set<Symbol> arguments = new HashSet<Symbol>();
```
Same as above

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+            Set<Symbol> arguments = new HashSet<Symbol>();
+            aggregations.values()
+                    .stream()
+                    .filter(functionCall -> functionCall.isDistinct())
```
FunctionCall::isDistinct

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+                            false,
+                            ImmutableList.of(argument.toSymbolReference())));
+                    functions.put(entry.getKey(),
+                            metadata.getFunctionRegistry().resolveFunction(functionName, ImmutableList.of(symbolAllocator.getTypes().get(argument).getTypeSignature()), false));
```
This is a little hard to read. I'd either introduce a variable for the parameter types or the function or format it across multiple lines.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+                }
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
```
First argument goes on the next line.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+        // creates if clause specific to use case here, default value always null
+        private SearchedCaseExpression createIfExpression(Expression left, Expression right, ComparisonExpression.Type type, Expression result)
+        {
+            return new SearchedCaseExpression(ImmutableList.of(createWhenClause(left, right, type, result)),
```
You can use IfExpression directly

stagraqubole commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+                    aggregations.put(entry.getKey(), new FunctionCall(functionCall.getName(),
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(Iterables.getOnlyElement(aggregateInfo.getNewDistinctAggregateSymbols().values()).toSymbolReference())));
```
Will change it, this was left as map  because I originally thought of writing this for multiple distinct aggregations.

stagraqubole commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,482 @@
+        {
+            this.groupBySymbols = ImmutableList.copyOf(groupBySymbols);
+
+            if (mask != null) {
```
right, this is unnecessary, will remove it.

stagraqubole commented:
Taken care of review comments @martint

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+                    Symbol newSymbol = symbolAllocator.newSymbol("expr", symbolAllocator.getTypes().get(symbol));
+                    aggregateInfo.setNewDistinctAggregateSymbol(newSymbol);
+
+                    Expression expression = createIfExpression(groupSymbol.toSymbolReference(),
```
This is probably ok for now, but it should eventually be replaced with `GROUPING` when it's available: https://github.com/prestodb/presto/pull/5465

martint replied:
First argument goes on next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+                else if (aggregationOuputSymbolsMap.containsKey(symbol)) {
+                    Symbol newSymbol = symbolAllocator.newSymbol("expr", symbolAllocator.getTypes().get(symbol));
+                    outputNonDistinctAggregateSymbols.put(aggregationOuputSymbolsMap.get(symbol), newSymbol); // key of this map is key of an aggregation in AggrNode above, it will now aggregate on this Map's value
+                    Expression expression = createIfExpression(groupSymbol.toSymbolReference(),
```
First argument goes on next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+
+            aggregateInfo.setNewNonDistinctAggregateSymbols(outputNonDistinctAggregateSymbols.build());
+
+            return new ProjectNode(idAllocator.getNextId(),
```
Place all arguments on the same line. It's not too long in this case.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+                    outputSymbols.build());
+        }
+
+        private GroupIdNode createGroupIdNode(List<Symbol> groupBySymbols,
```
First argument goes on next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+         * The new AggregateNode aggregates on the symbols that original AggregationNode aggregated on
+         * Original one will now aggregate on the output symbols of this new node
+         */
+        private AggregationNode createNonDistinctAggregation(AggregateInfo aggregateInfo,
```
First argument goes on next line

stagraqubole commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+ *  INTO
+ *  SELECT a1, a2,..., an, arbitrary(f1 if group = 0 else null),...., arbitrary(fm if group = 0 else null), F(c if group = 1 else null) FROM
+ *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c, group FROM
+ *        SELECT a1, a2,..., an, b1, b2, ... ,bn, c FROM Table GROUP BY GROUPING SETS ((a1, a2,..., an, b1, b2, ... ,bn), (a1, a2,..., an, c))
```
Yes, it can happen but only if there are some aggregate functions which can work on HyperLogLog. There are no such functions today, not sure if aggregate functions on HyperLogLog are useful. Should I just add the check to bail out for non-equatable types to be on safer side?

stagraqubole commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java:
```diff
@@ -0,0 +1,481 @@
+                    }
+                    // Cast to avoid UnaliasSymbolReference from removing this projection
+                    // TODO: find better way to prevent UnaliaSymbolRefrence optimizing out the project
+                    builder.put(newSymbol, new Cast(distinctSymbol.toSymbolReference(),
```
Let it be here? Or wait for #6379 to be fixed?

martint replied:
It's a correctness issue, which we treat with very high priority. We'll fix it shortly.

stagraqubole replied:
Ok, I will update the PR once it is fixed

yuananf commented:
@stagraqubole will this kind of query be optimized?
```
presto:tpcds_bin_partitioned_orc_1000> select count(distinct ss_sold_time_sk), count(distinct ss_item_sk) from store_sales;
 _col0 | _col1  
-------+--------
 46800 | 300000 
(1 row)

Query 20161024_075438_00027_4gzht, FINISHED, 10 nodes
Splits: 1,905 total, 1,905 done (100.00%)
2:14 [2.88B rows, 9.84GB] [21.5M rows/s, 75.1MB/s]
```
It finished in over 2 minutes, but the following queries finished in less than 10 seconds.

```
presto:tpcds_bin_partitioned_orc_1000> select count(distinct ss_sold_time_sk) from store_sales;
 _col0 
-------
 46800 
(1 row)

Query 20161024_075416_00025_4gzht, FINISHED, 10 nodes
Splits: 1,895 total, 1,895 done (100.00%)
0:03 [2.88B rows, 2.76GB] [827M rows/s, 813MB/s]

presto:tpcds_bin_partitioned_orc_1000> select count(distinct ss_item_sk) from store_sales;
 _col0  
--------
 300000 
(1 row)

Query 20161024_075424_00026_4gzht, FINISHED, 10 nodes
Splits: 1,895 total, 1,895 done (100.00%)
0:06 [2.88B rows, 8.5GB] [449M rows/s, 1.33GB/s]

```

stagraqubole commented:
@yuananf current PR will not optimize the query you mentioned. Right now, queries with single distinct mask and multiple aggregations will be optimized.
The plan is to extend these changes to handle multiple distinct masks as well. I will be working on it once this PR is approved. With those changes, the query you mentioned will also be optimized

yuananf commented:
@stagraqubole thanks, that would be great.

stagraqubole commented:
@martint took care of review comments, back to you

martint commented:
Merged (with a few minor changes). Thanks!

BTW, I renamed the optimizer, config and session properties.
